### ISSUE TITLE
(paddlepaddle+paddleocr) Running standalone compiled exe leads to libpaddle.pyp not being recognized as a package

### ISSUE THREAD

**Issue by sreid19** [Nuitka/Nuitka#2982]
Version: 2.3.11
Commercial: None
Python: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:03:56) [MSC v.1929 64 bit (AMD64)]
Flavor: Anaconda Python
Executable: C:\Users\scott\anaconda3\envs\exe_only\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 11

- Installed using pip into an anaconda environment (tried the development branch too). I also tried it with a venv and using pip to install the packages. I used tried the anaconda c compiler, the default compiler when using venv and python 3.12.4, and also clang for both anaconda and venv environments and both gave the error. I also tried using python 3.11.

- packages
paddleocr              2.7.3      
paddlepaddle           2.6.1       

- To reproduce:
pip install paddlepaddle
pip install paddleocr

in a .py file named startup.py:
from paddleocr import PaddleOCR

python -m nuitka startup.py --standalone --enable-plugin=tk-inter

C:\Users\scott\website\exe_only\cod-site\startup.dist>startup.exe
Traceback (most recent call last):
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\startup.py", line 21, in <module>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddleocr\__init__.py", line 14, in <module paddleocr>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddleocr\paddleocr.py", line 21, in <module paddleocr.paddleocr>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddle\__init__.py", line 28, in <module paddle>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddle\base\__init__.py", line 39, in <module paddle.base>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddle\base\framework.py", line 37, in <module paddle.base.framework>
  File "C:\Users\scott\website\exe_only\cod-site\STARTU~1.DIS\paddle\pir\__init__.py", line 15, in <module paddle.pir>
ModuleNotFoundError: No module named 'paddle.base.libpaddle.pir'; 'paddle.base.libpaddle' is not a package

- If I do not use standalone at all, the exe starts up just fine. Using --follow-imports instead of --standalone produces the same issue as standalone.

- This goes without saying, but my code runs fine in my python environment. Don't think it's a path issue as it is finding the .pyp file, but for some reason it isn't recognizing it as a package.

**Comment by kayhayen** [comment]
This might be a python 3.12 only issue. Extension modules no longer can be informed what package to load into. This will only be solved much later for windows. For Linux there will be fix with static libpython only. Workarounds will take time

**Comment by sreid19** [comment]
Interesting, when using both python 3.10 and 3.11 it crashes with FATAL: anti-bloat: Making changes to 'paddleocr.paddleocr' that cause SyntaxError 'unmatched ')' (paddleocr.py, line 48)' in the first pass.

**Comment by kayhayen** [comment]
Oh, we might have a mistake there. Versions are the same though, because that would be surprising. If not, please share the version you used with me here, so I can look into fixing that issue.

**Comment by sreid19** [comment]
Here is my version info for when this bug appears:
2.3.11
Commercial: None
Python: 3.10.0 | packaged by conda-forge | (default, Nov 10 2021, 13:20:59) [MSC v.1916 64 bit (AMD64)]
Flavor: Anaconda Python
Executable: C:\Users\scott\anaconda3\envs\exe\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 11

I also tried with version 2.0.4 as I saw another thread saying there was a fix for paddleocr in that version.

**Comment by sreid19** [comment]
I was able to temporarily solve the issue by reverting to paddleocr 2.7.3. I now unfortunately get a module not found error with django.views.debug. I will try some workarounds using --include-package data, and if I cannot solve it I will open a separate issue.

**Comment by kayhayen** [comment]
For Django dependencies I have done stuff, that's probably on develop or factory branches right now, and it will be in 2.4 release really soon. I am not aware of a need to package data for `django.views.debug` and it might be hard to use with Nuitka, depending on what it wants to do. If you can, make a separate issue with a reproducer for it. I am in the process of adding full Django out of the box support. Right now, installed apps are automatically added, soon database will be done too, and I am interested in reproducers here, well, I mean separate issues.

**Comment by kayhayen** [comment]
Ok, I got it. We need ideas @KRRT7 for code like this:

```
- module-name: 'paddleocr.paddleocr' # checksum: 8cf243fc
  anti-bloat:
    - replacements_plain:
        'tools = _import_file(': ''
        "    'tools', os.path.join(__dir__, 'tools/__init__.py'), make_importable=True)": 'from . import tools'
```

They changed the quoting style there, which is why it's broken. It's now like this:

> "tools", os.path.join(__dir__, "tools/__init__.py"), make_importable=True

This is way too fragile. I think a regular expression might be better to use here, but yeah, without a proper monitoring of it breaks, these things are bound to regress any day.

**Comment by kayhayen** [comment]
Ah, they also changed the whitespace for the closing brace, yeah, regular expressions are the only way to cover both.

**Comment by kayhayen** [comment]
So far the replacements_re is constrained to multiline matches I think, needed to first change that, or else it can also only work on individual lines. For replace source code, I feel ideally we would have stuff that makes matching for white space and quoting styles not matter, but I have no idea how to do that. It would probably have to match on ast.level and that sounds horribly hard.

**Comment by KRRT7** [comment]
> Ok, I got it. We need ideas @KRRT7 for code like this:
> 
> ```
> - module-name: 'paddleocr.paddleocr' # checksum: 8cf243fc
>   anti-bloat:
>     - replacements_plain:
>         'tools = _import_file(': ''
>         "    'tools', os.path.join(__dir__, 'tools/__init__.py'), make_importable=True)": 'from . import tools'
> ```
> 
> They changed the quoting style there, which is why it's broken. It's now like this:
> 
> > "tools", os.path.join(**dir**, "tools/**init**.py"), make_importable=True
> 
> This is way too fragile. I think a regular expression might be better to use here, but yeah, without a proper monitoring of it breaks, these things are bound to regress any day.

running something like black on source code we might be replacing, might work well. if the code is formatted then the target replacement will be the same regardless of what they do

**Comment by sreid19** [comment]
> For Django dependencies I have done stuff, that's probably on develop or factory branches right now, and it will be in 2.4 release really soon. I am not aware of a need to package data for `django.views.debug` and it might be hard to use with Nuitka, depending on what it wants to do. If you can, make a separate issue with a reproducer for it. I am in the process of adding full Django out of the box support. Right now, installed apps are automatically added, soon database will be done too, and I am interested in reproducers here, well, I mean separate issues.

Turns out the issue was a conflict with asgiref, which I solved by downgrading to 3.7.2

**Comment by kayhayen** [comment]
Either it was not real, but it definitely seems like the reporter lost interest, I don't see this progressing without a new report.
